

Дано 4 точки: $M_0(0, 0, 0), M_1(1, \frac{1}{3}, 0), M_2(0, 2, \frac{1}{4}), M_3(\frac{1}{2}, \frac{1}{2}, 1)$.

Також дано 3 вектори, які задають паралелепіпед із зазначеними точками:
- $\overline{a}=\overline{M_0M_1}$
- $\overline{b}=\overline{M_0M_2}$
- $\overline{c}=\overline{M_0M_3}$

Порахуй координати векторів через задані точки.  
Порахуй об'єм, площу повної поверхні паралелепіпеда та кути між ребрами паралелепіпеда, координати решти вершин паралелепіпеда.



In [4]:
import plotly.graph_objects as go
import numpy as np

def parallelepiped_volume(a, b, c):
    mixed = np.linalg.det(np.dstack([a,b,c]))
    return np.linalg.norm(mixed)

def parallelepiped_surface_area(a, b, c):
    cross_product_ab = np.cross(a,b)
    cross_product_bc = np.cross(b,c)
    cross_product_ca = np.cross(c,a)
    side_area_ab = np.linalg.norm(cross_product_ab)
    side_area_bc = np.linalg.norm(cross_product_bc)
    side_area_ca = np.linalg.norm(cross_product_ca)
    total_surface_area = 2 * (side_area_ab + side_area_bc + side_area_ca)
    return total_surface_area

def calculate_cosine_angle(vector1, vector2):
    dot_product = np.dot(vector1, vector2)
    magnitude_vector1 = np.linalg.norm(vector1)
    magnitude_vector2 = np.linalg.norm(vector2)

    cosine_angle = np.arccos(dot_product / (magnitude_vector1 * magnitude_vector2))

    return np.degrees(cosine_angle)

vector_a = np.array([1, 1/3, 0])
vector_b = np.array([0, 2, 1/4])
vector_c = np.array([1/2, 1/2, 1])
vector_ac = vector_a + vector_c
vector_bc = vector_b + vector_c
vector_ab = vector_a + vector_b
vector_abc = vector_a + vector_b + vector_c

fig = go.Figure()
fig.add_trace(go.Scatter3d(x=[0, vector_a[0]], y=[0, vector_a[1]], z=[0, vector_a[2]],
                           mode='lines+markers', marker=dict(size=5), line=dict(color='blue'), name='Вектор A(M0M1)',))
fig.add_trace(go.Scatter3d(x=[0, vector_b[0]], y=[0, vector_b[1]], z=[0, vector_b[2]],
                           mode='lines+markers', marker=dict(size=5), line=dict(color='red'), name='Вектор B(M0M2)'))
fig.add_trace(go.Scatter3d(x=[0, vector_c[0]], y=[0, vector_c[1]], z=[0, vector_c[2]],
                           mode='lines+markers', marker=dict(size=5), line=dict(color='green'), name='Вектор C(M0M3)'))
fig.add_trace(go.Scatter3d(x=[vector_a[0], vector_ac[0]], y=[vector_a[1], vector_ac[1]], z=[vector_a[2], vector_ac[2]],
                           mode='lines', line=dict(color='black', dash='dash'), name='Соединительная линия'))
fig.add_trace(go.Scatter3d(x=[vector_c[0], vector_ac[0]], y=[vector_c[1], vector_ac[1]], z=[vector_c[2], vector_ac[2]],
                           mode='lines', line=dict(color='black', dash='dash'), name='Соединительная линия'))
fig.add_trace(go.Scatter3d(x=[vector_b[0], vector_bc[0]], y=[vector_b[1], vector_bc[1]], z=[vector_b[2], vector_bc[2]],
                           mode='lines', line=dict(color='black', dash='dash'), name='Соединительная линия'))
fig.add_trace(go.Scatter3d(x=[vector_c[0], vector_bc[0]], y=[vector_c[1], vector_bc[1]], z=[vector_c[2], vector_bc[2]],
                           mode='lines', line=dict(color='black', dash='dash'), name='Соединительная линия'))
fig.add_trace(go.Scatter3d(x=[vector_a[0], vector_ab[0]], y=[vector_a[1], vector_ab[1]], z=[vector_a[2], vector_ab[2]],
                           mode='lines', line=dict(color='black', dash='dash'), name='Соединительная линия'))
fig.add_trace(go.Scatter3d(x=[vector_b[0], vector_ab[0]], y=[vector_b[1], vector_ab[1]], z=[vector_b[2], vector_ab[2]],
                           mode='lines', line=dict(color='black', dash='dash'), name='Соединительная линия'))
fig.add_trace(go.Scatter3d(x=[vector_ab[0], vector_abc[0]], y=[vector_ab[1], vector_abc[1]], z=[vector_ab[2], vector_abc[2]],
                           mode='lines', line=dict(color='black', dash='dash'), name='Соединительная линия'))
fig.add_trace(go.Scatter3d(x=[vector_ac[0], vector_abc[0]], y=[vector_ac[1], vector_abc[1]], z=[vector_ac[2], vector_abc[2]],
                           mode='lines', line=dict(color='black', dash='dash'), name='Соединительная линия'))
fig.add_trace(go.Scatter3d(x=[vector_bc[0], vector_abc[0]], y=[vector_bc[1], vector_abc[1]], z=[vector_bc[2], vector_abc[2]],
                           mode='lines', line=dict(color='black', dash='dash'), name='Соединительная линия'))


fig.add_trace(go.Scatter3d(x=[vector_ac[0]], y=[vector_ac[1]], z=[vector_ac[2]], mode='markers+text', marker=dict(size=5), text=['M4 (1.5, 0.83, 1)'], textposition='top center',  name='Точка M4'))
fig.add_trace(go.Scatter3d(x=[vector_bc[0]], y=[vector_bc[1]], z=[vector_bc[2]], mode='markers+text', marker=dict(size=5), text=['M5 (0.5, 2.5, 1.25)'], textposition='top center',  name='Точка M5'))
fig.add_trace(go.Scatter3d(x=[vector_ab[0]], y=[vector_ab[1]], z=[vector_ab[2]], mode='markers+text', marker=dict(size=5), text=['M6 (1, 2.3, 0.25)'], textposition='top center',  name='Точка M6'))
fig.add_trace(go.Scatter3d(x=[vector_abc[0]], y=[vector_abc[1]], z=[vector_abc[2]], mode='markers+text', marker=dict(size=5), text=['M7 ([1.5, 2.8, 1.25)'], textposition='top center',  name='Точка M7'))




fig.update_layout(scene=dict(aspectmode='data'))
fig.update_layout(scene=dict(xaxis_title='Вісь X', yaxis_title='Вісь Y', zaxis_title='Вісь Z'))
fig.show()

V = parallelepiped_volume(vector_a, vector_b, vector_c)
S = parallelepiped_surface_area(vector_a, vector_b, vector_c)
angle_ab =  calculate_cosine_angle(vector_a, vector_b)
angle_ac =  calculate_cosine_angle(vector_a, vector_c)
angle_bc =  calculate_cosine_angle(vector_b, vector_c)
angle_ab, angle_ac, angle_bc

(71.71246981426138, 58.90906964232697, 59.57768271314148)

$\overline{a} = M_1-M_0 = \left (\begin{array}{cc}
1\\
\frac{1}{3}\\
0
\end{array}\right)
\\
\overline{b} = M_2-M_0 = \left (\begin{array}{cc}
0\\
2\\
0.25
\end{array}\right)
\\
\overline{c} = M_3-M_0 = \left (\begin{array}{cc}
0.5\\
0.5\\
1
\end{array}\right)
$

$
V =|\overline{a} * (\overline{b} * \overline{c})| = 1.91\\
S = 10.5\\
\angle(a, b) = 71.7\\
\angle(a, c) = 58.9\\
\angle(b, c) = 59.6\\
M_4 = (1.5, 0.83, 1)\\
M_5 = (0.5, 2.5 , 1.25)\\
M_6 = (1  , 2.3 , 0.25)\\
M_7 = (1.5, 2.83, 1.25)
$